database/database.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		err = fmt.Errorf("failed to set the max. number of DBs to 1: "+
+			"%s", err)
+		closeErr := env.Close()
+		if closeErr != nil {
+			err = fmt.Errorf("%s; failed to close the database: "+
+				"%s", err, closeErr)
+		}
+		return
+	}
+
+	mapSize := int64(32 * 1024 * 1024 * 1024)
+	err = env.SetMapSize(mapSize)
+	if err != nil {
+		err = fmt.Errorf("failed to set the map size to "+
+			"%d: %s", mapSize, err)
+		closeErr := env.Close()
+		if closeErr != nil {
+			err = fmt.Errorf("%s; failed to close the "+
+				"database: %s", err, closeErr)
+		}
+		return
+	}
+
+	return
+}
+
+// Database represents a database of channels.
+type Database struct {
+	Path     string
+	env      *lmdb.Env
+	Readonly bool
+}
+
+// Initialize initializes the database environment in the given directory.
+// Creates the expected database and should be called only once during the
+// deployment.
+func Initialize(path string) (err error) {
+	_, err = os.Stat(path)
+	if err != nil {
+		err = fmt.Errorf("the directory of LMDB environment "+
+			"is expected to exist: %s", err.Error())
+		return
+	}
+
+	env, err := newEnv()
+	if err != nil {
+		return
+	}
+	defer func() {
+		err = env.Close()
+	}()
+
+	err = env.Open(path, 0, 0660)
+	if err != nil {
+		return
+	}
+
+	err = env.Update(func(txn *lmdb.Txn) (txnErr error) {
+		_, txnErr = txn.OpenDBI(dbChannelName, lmdb.Create)
+		if txnErr != nil {
+			return
+		}
+		return
+	})
+
+	return
+}
+
+// NewDatabase creates a new database object. The database directory is assumed
+// to be already initialized.
+func NewDatabase(
+	path string,
+	readonly bool) (d *Database, err error) {
+
+	env, err := newEnv()
+	if err != nil {
+		return
+	}
+
+	if readonly {
+		err = env.Open(path, lmdb.Readonly, 0440)
+	} else {
+		err = env.Open(path, 0, 0660)
+	}
+
+	if err != nil {
+		err = fmt.Errorf("failed to open the database in the "+
+			"directory %s: %s", path, err)
+		closeErr := env.Close()
+		if closeErr != nil {
+			err = fmt.Errorf("%s; failed to close the database: "+
+				"%s", err, closeErr)
+		}
+		return
+	}
+
+	d = &Database{
+		Path:     path,
+		env:      env,
+		Readonly: readonly}
+
+	return
+}
+
+// GetChannel returns the channel associated with the descriptor in the
+// database, if it exists; nil otherwise.
+func (t *Txn) GetChannel(descriptor string) (channel *protoed.Channel,
+	err error) {
+	encodedKey := Descriptor(descriptor).Encode()
+
+	value, getErr := t.lmdbTxn.Get(t.channelDbi, encodedKey)
+	switch {
+	case getErr == nil:
+		// pass
+	case lmdb.IsNotFound(getErr):
+		// not found, return
+		return
+	default:
+		err = fmt.Errorf("failed to get the channel: %s", getErr.Error())
+		return
+	}
+
+	channel = &protoed.Channel{}
+	err = proto.Unmarshal(value, channel)
+
+	if err != nil {
+		err = fmt.Errorf("failed to unmarshal the channel: %s",
+			err.Error())
+		return
+	}
+
+	return
+}
+
+// Count returns the number of items in the database.
+func (t *Txn) Count() (count uint64, err error) {
```

Comment:
While I can't add the postcondition you mention (we'd need logical quantifiers forall and exists which are not supported by gocontracts), I can add the postcondition to GetChannel that if error is nil and channel is not nil, count is positive.

Reply:
I was more thinking of what happens when you `Get` on an empty database -- an error or a nil result?

Reply:
that case is covered in the tests. Upon `Get`ting a non-existent element, the error `lmdb.IsNotFound` is thrown, which we catch and ignore;  since `Count` is computed through the `Stat` method of the lmdb, it returns an errorless 0 for an empty database.

Reply:
I would also add a contract as formal documentation -- Get ensures that, if in testing and MustCount is 0, that a nil result and a nil error is returned.